### ISSUE TITLE
Diffusion Models Beat GANs on Image Synthesis

### ISSUE THREAD

**Issue by reyllama** [reyllama/paper-reviews#159]
## TL; DR

- With classifier guidance and technical modifications to the model architecture, denoising diffusion probabilistic models can be trained to beat SOTA GANs on image generation benchmarks.

## Motivation

- Previous works on diffusion models brought remarkable advancements in this relatively new field of generative models, but there has been a shortage of study on specific ways to improve performance compared to GANs.

## Prior Works

- DDPM: draws connection between diffusion models and denoising score matching (Langevin sampling as well)
- DDIM: greatly reduces sampling time using theoretical derivation of non-Markovian process with matching marginals and training objective
- Improved DDPM: introduce technical hacks to improve log-likelihood and sample quality

### Architecture Improvements

![image](https://user-images.githubusercontent.com/46308407/153364835-039e0082-6acc-4cdb-93a5-4814b3cc718e.png)
![image](https://user-images.githubusercontent.com/46308407/153364863-c42fbb83-00a8-409d-b94b-fd50c2d675cd.png)
![image](https://user-images.githubusercontent.com/46308407/153364894-b3568b0c-ed6e-495b-86ad-79fdaad08d39.png)

- Final version: 2 resblocks / resolution (in UNet), multihead attention blocks w/ 64 channels, multi-resolution attention at 32, 16, 8, BigGAN resblocks and adaptive group normalization for injecting learned timestep and class embeddings to the residual blocks.

### Classifier Guidance

- Class-conditioning can provide much information for conditional generation tasks (class-conditioned normalization is widely used for GANs along with projection discriminator)

- Using Bayes rule, we can derive class conditional denoising transition as follows:
![image](https://user-images.githubusercontent.com/46308407/153365419-49d027f2-1942-44a2-aca5-2e56b42e718d.png)

- Decomposing class-conditioned reverse process into product of two densities, we simplify each.
- Simple reverse transition is modeled with learned parametric Gaussian distribution. (mu is typically modeled with a neural network while sigma is given as constant for simplification)

![image](https://user-images.githubusercontent.com/46308407/153365623-883880b5-71ea-4efa-96ac-e5bf130b25b6.png)

- Applying Taylor expansion to classifier distribution, we obtain something like: (g is gradient of log classifier distribution)
![image](https://user-images.githubusercontent.com/46308407/153365729-04d8c073-a21c-496a-9bfd-3c57a53cdb1b.png)

- Combining these two, we get
![image](https://user-images.githubusercontent.com/46308407/153365762-e67e4ff4-7042-4f7c-bfb4-f2e0231d7906.png)

- Unlike DDPM that relies on stochastic forward and backward process, DDIM makes the mapping deterministic by setting \beta=0.
- To incorporate classifier guidance on DDIM, we employ score-based conditioning trick.

- First, derive a score function from our prediction on the noise at timestep t.
![image](https://user-images.githubusercontent.com/46308407/153366213-8230fd75-7d4c-4df8-bbc6-e936768276d6.png)
![image](https://user-images.githubusercontent.com/46308407/153366273-681146c8-f07d-4959-b192-9a188f0d936f.png)

![image](https://user-images.githubusercontent.com/46308407/153366295-e7481d1f-5625-4d5f-aba7-62eab3ddc5f1.png)

### Gradient Scales

- With gradient scale s=1 (default), the sample quality is not satisfactory.
- With larger gradient scale, the classifier output distribution (post-softmax) is sharpened (equivalent to smaller softmax temperature)
- Hence, the gradient more effectively guide diffusion process towards the class manifold at the cost of diversity.
- Just like truncation tricks widely used in GAN literature, we can trade diversity off for precision in diffusion models as well.

![image](https://user-images.githubusercontent.com/46308407/153366689-5231c373-0df1-410d-b37b-bc9c1670bb1d.png)

### Effect of gradient scale (precision-recall trade-off)

![image](https://user-images.githubusercontent.com/46308407/153366735-3137976a-b3fc-43ee-a52f-856ea5c5f469.png)
![image](https://user-images.githubusercontent.com/46308407/153366765-6ff6fa01-bf67-4985-b48d-2185849dc02d.png)

### State-of-the-Art on various benchmarks

![image](https://user-images.githubusercontent.com/46308407/153367010-7fc4f366-1e6a-4c5a-9833-d7050b3ebddf.png)
![image](https://user-images.githubusercontent.com/46308407/153367048-60348741-ad13-4320-8f24-295b32f7f696.png)

### Classifier guidance and upsampling

![image](https://user-images.githubusercontent.com/46308407/153367116-eb2ba51d-6521-4081-9c3f-4ebc526759ff.png)

## Comments

- It's very elegant how techniques like classifier guidance can be derived theoretically.
- Although the advent of a new family of generative models is very exciting, quantitatively superior results is at best *necessary* to compensate for many inconveniences (additional classifier training, slow sampling)
- The amount of work put into diffusion models is extremely small compared to GANs, which has been at the center for attention in generative models community for nearly a decade (8 years?)
- So, I think there is much room for improvements. Some of the improvements could include getting rid of additional classifier, further accelerating sampling, producing images of higher resolution (1024 or more), incorporating transformer, etc.
